### DISCUSSION TITLE
What is the best way to include page content using a template file in order

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3376] [Q&A]
Currently have a setup where I pull info from a list inside a page metadata to create a timeline to display (Result can be seen [here](https://discord.hasgoodui.design)).

While this setup works do I see 2 essential issues with it:

1. It fills the metadata, making it unreasonably large
2. Content is limited to that particular page, making things like category pages not possible.

I now consider making a small MkDocs plugin which would pull files from a folder, sort them by date and then replace a specific placeholder with the content.
If possible would I also like to use a template file, so that I could just change that instead of having to update the plugins code later on...

The question I have now is: What would be the best way to achieve this? The closest experience I have is a page-nav generator, which creates unordered lists of sub-pages of one page... "experience" is a word I would use lightly tho, as that one was effectively a copy-paste I got from the awesome-pages plugin dev (Thanks for that. Still appreciate the help you gave).

So I was looking at other plugins like the neotori.timeline one, which my setup is inspired by (And shamelessly copying CSS setups from) for how they do some stuff and the material for mkdocs blog plugin to *hopefully* get an understanding on how to retrieve posts, sort them etc.

I know I ask for a lot, but I'm sadly a "learning by doing" person who looks at examples to learn from, but right now am I at a dead end to understand things...

My ideal goal would be to make a plugin which has a placeholder with optional category to go by (Set inside metdata of a page. Multiple ones should idealy be doable), CSS for the styling (That a possible user could override) and aforementioned template file to inject content into.

I appreciate any hints into the right direction here. The closest I have in terms of any code is... this, which is not a lot to say the least.
```python
import re

from mkdocs.plugins import BasePlugin

class MkDocsTimelineGeneratorPlugin(BasePlugin):
  rx = re.compile(r"{timeline(?::([^\)]))?}")

  def on_post_page(self, output, page, config):
    if "{timeline" in output:
      """No idea what to do here... :sweat_smile:"""

  def _render_content(self, match) -> str:
    """Would this even work?"""
    source = "__all__" if not match.group(1) else match.group(1)
```

**Comment by pawamoy** [comment]
If you take a look at my "website" sources, you'll see that I override templates and add new ones: https://github.com/pawamoy/website/tree/main/docs/templates. Each Markdown page can then tell which template to use:

- index with my posts list: https://raw.githubusercontent.com/pawamoy/website/main/docs/index.md
- showcase, with data directly in front matter too: https://raw.githubusercontent.com/pawamoy/website/main/docs/showcase.md
- etc.

MkDocs knows where to find my templates with this:

```yaml
theme:
  name: material
  custom_dir: docs/templates
```

That's all I can give right now :sweat_smile: I think it's not so far from what you're doing, and shows how to use templates.

**Reply by Andre601** [reply]
Thanks for the info.

Tho, I noticed one issue... Your `post_list.html` seems to use `{{ page.meta.data.strftime("%b %d, %Y") }}` but when I try to use that to format time do I only get a `'str object' has no attribute 'strftime'`

Do you add/use anything specific here to make this work?

**Reply by Andre601** [reply]
Another small inconvenience I currently find is, that I cannot sort the pages by their date, starting with the newest... I refer to the `date` frontmatter.

Any ideas?

**Reply by pawamoy** [reply]
> `{{ page.meta.data.strftime("%b %d, %Y") }}`

Typo maybe? You wrote data here instead of date. Otherwise maybe the Material for MkDocs theme transforms it into a date, explaining why I can use strftime on it. Do you use another theme?

Dates or strings, sorting should be natively supported. You can use the [`sort`](https://jinja.palletsprojects.com/en/3.0.x/templates/#jinja-filters.sort) Jinja filter:

```jinja+html
{% for page in nav.pages|sort(attribute="meta.date", reverse=True) %}
  ...
{% endfor %}
```

**Reply by Andre601** [reply]
It's a typo here. I did use `date` in the actual source. And yes, Material for MkDocs is used here, so the date is automatically translated.

The problem with sort is, that it doesn't work. See #3377 where I tried using `meta.date` but it gave me an `UndefinedError` which is confusing to say the least. I also mentioned that no matter what I tried to add, the sort option has no effect on the result...

At one point did I manage to get sort to recognize the date attribute, but it gave me an error then because the value was not a number apparently, so I first would need to try and transform it into a number?

It's really confusing and I wish MkDocs had some way to sort nav by values like this.

**Reply by pawamoy** [reply]
Your template will only impact the pages that use it, the actual nav will stay untouched. Not sure what you mean by "result" here. We'd need a more concrete example: docs layout and files, what you do with them, and the layout you expect.

**Reply by Andre601** [reply]
I just want to create a timeline of entries based on what is in a specific folder. And have this timeline sorted by the date of each post.

I've already linked how it looks like in my initial post, but here's the link again just in case: https://discord.hasgoodui.design
